google/resource_compute_disk.go

```diff
@@ -390,213 +760,377 @@ func resourceComputeDiskDelete(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*Config)
 
-		// Translate edition
-		switch edition {
-		case "enterprise":
-			edition = "ent"
-		case "standard":
-			edition = "std"
-		case "express":
-			edition = "exp"
+	project, err := getProject(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the zone
+	z, err := getZone(d, config)
+	if err != nil {
+		return nil, err
+	}
+
+	zone, err := config.clientCompute.Zones.Get(project, z).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	if v, ok := d.GetOk("image"); ok {
+		log.Printf("[DEBUG] Resolving image name: %s", v.(string))
+		imageUrl, err := resolveImage(config, project, v.(string))
+		if err != nil {
+			return nil, fmt.Errorf(
+				"Error resolving image name '%s': %s",
+				v.(string), err)
 		}
 
-		var f string
-		if revision := parts[4]; revision != "" {
-			// With revision
-			f = fmt.Sprintf("sql-%s-%s-win-%s-r%s", edition, sqlVersion, windowsVersion, revision)
-		} else {
-			// No revision
-			f = fmt.Sprintf("sql-%s-%s-win-%s", edition, sqlVersion, windowsVersion)
+		obj["sourceImage"] = imageUrl
+		log.Printf("[DEBUG] Image name resolved to: %s", imageUrl)
+	}
+
+	if v, ok := d.GetOk("type"); ok {
+		log.Printf("[DEBUG] Loading disk type: %s", v.(string))
+		diskType, err := readDiskType(config, zone, project, v.(string))
+		if err != nil {
+			return nil, fmt.Errorf(
+				"Error loading disk type '%s': %s",
+				v.(string), err)
 		}
 
-		if f == familyName {
-			return true
+		obj["type"] = diskType.SelfLink
+	}
+
+	if v, ok := d.GetOk("snapshot"); ok {
+		snapshotName := v.(string)
+		match, _ := regexp.MatchString("^https://www.googleapis.com/compute", snapshotName)
+		if match {
+			obj["sourceSnapshot"] = snapshotName
+		} else {
+			log.Printf("[DEBUG] Loading snapshot: %s", snapshotName)
+			snapshotData, err := config.clientCompute.Snapshots.Get(
+				project, snapshotName).Do()
+
+			if err != nil {
+				return nil, fmt.Errorf(
+					"Error loading snapshot '%s': %s",
+					snapshotName, err)
+			}
+			obj["sourceSnapshot"] = snapshotData.SelfLink
 		}
 	}
 
-	return false
+	return obj, nil
 }
 
-// e.g. image: windows-server-1709-dc-core-v20180109, family: windows-1709-core
-// e.g. image: windows-server-1709-dc-core-for-containers-v20180109, family: "windows-1709-core-for-containers
-func suppressWindowsFamilyDiff(imageName, familyName string) bool {
-	updatedFamilyString := strings.Replace(familyName, "windows-", "windows-server-", 1)
-	updatedFamilyString = strings.Replace(updatedFamilyString, "-core", "-dc-core", 1)
+func resourceComputeDiskDecoder(d *schema.ResourceData, meta interface{}, res map[string]interface{}) (map[string]interface{}, error) {
+	if v, ok := res["diskEncryptionKey"]; ok {
+		original := v.(map[string]interface{})
+		transformed := make(map[string]interface{})
+		// The raw key won't be returned, so we need to use the original.
+		transformed["raw_key"] = d.Get("disk_encryption_key.0.raw_key")
```

Comment:
I think these have to be `rawKey` to match what the flatteners are expecting.

On another note, it would be great to have tests for some of these trickier spots.

Reply:
Shoot, forgot that the flatteners get called here.  Writing more, better tests.